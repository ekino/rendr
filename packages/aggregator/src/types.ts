import { BlockDefinition, Page, RendrCtx } from "@ekino/rendr-core";

export type HandlerList = {
  [index: string]: Handler;
};

export type Handler = (
  definition: BlockDefinition,
  ctx: RendrCtx,
  page: Page
) => Promise<BlockDefinition>;

export type HandlerRegistry = (code: string) => Handler;
