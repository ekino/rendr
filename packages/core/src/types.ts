import { IncomingMessage, ServerResponse } from "http";

export interface BlockDefinition {
  type: string;
  settings: Settings;
  container: string;
  order: number;
}
export interface Settings {
  [index: string]: any;
}

export interface Cache {
  ttl: number;
}

export interface Head {
  titleTemplate: string;
  defaultTitle: string;
  title: string;
  link: string;
  htmlAttributes: Settings;
  meta: Settings[];
}

// -- Page definition
export class Page {
  public statusCode: number = 200;
  public type: string = "document";
  public template: string = "default";
  public cache: Cache = {
    ttl: 0
  };
  public head: Head = {
    titleTemplate: "%s",
    defaultTitle: "-",
    title: "-",
    link: "",
    htmlAttributes: {},
    meta: []
  };
  public blocks: BlockDefinition[] = [];
  public settings: Settings[] = [];
  public id: string = "";
}

// -- NextJs Controller signature
export interface RequestCtx {
  pathname: string;
  params: {};
  query: string;
  asPath: string;
  req: IncomingMessage;
  res: ServerResponse;
}
