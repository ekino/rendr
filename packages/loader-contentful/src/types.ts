import {
  Asset as ContentfulAsset,
  Entry,
  ContentfulCollection
} from "contentful";
import { RequestCtx, BlockDefinition } from "@ekino/rendr-core";

export { Asset as ContentfulAsset } from "contentful";

import { ContentfulClientApi } from "contentful";

export interface EntryNormalizer {
  (entry: Entry<any> | ContentfulAsset, normalizers?: EntryNormalizer):
    | any
    | void;
}

export type EntryNormalizerList = {
  [index: string]: EntryNormalizer;
};

export interface ClientFactory {
  (ctx: RequestCtx): ContentfulClientApi;
}

export interface Website {
  id: string;
  name: string;
  domains: string[];
  path: string;
  culture: string;
  countryCode: string;
  order: number;
  enabled: boolean;
}

export interface Author {
  id: string;
  name: string;
  jobTitle: string;
  slug: string;
  social: {
    twitter: string;
    facebook: string;
    linkedin: string;
  };
  biography: string;
  image: Asset;
}

export interface Asset {
  url: string;
  title: string;
  id: string;
}

export interface Article {
  id: string;
  type: string;
  title: string;
  abstract: string;
  seo: {
    description: string;
    keywords: string;
  };
  slug: string;
  website: Website;
  published_at: string;
  authors: Author[];
  blocks: BlockDefinition[];
  images: {
    list: Asset;
    header: Asset;
  };
}

export interface ContentfulWebsite {
  name: string;
  domains: string[];
  path: string;
  culture: string;
  country_code: string;
  order: number;
  enabled: boolean;
}

export interface ContentfulAuthor {
  name: string;
  job_title: string;
  slug: string;
  social_twitter: string;
  social_facebook: string;
  social_linkedin: string;
  biography: string;
  image: ContentfulAsset;
}

export interface ContentfulPage {
  title: string;
  type: string;
  seo_description: string;
  seo_keywords: string;
  extends: string;
  path: string;
  website: ContentfulWebsite;
  layout: string;
  ttl: number;
  settings: object;
  published_at: string;
  blocks: Entry<any>[];
  code: string;
}

export interface BaseContentfulBlock {
  internal_title: string;
  container: string;
  order: number;
}

export interface ContentfulBlockText extends BaseContentfulBlock {
  title: string;
  subtitle: string;
  contents: string;
  mode: string;
  image: ContentfulAsset;
  image_position: string;
}

export interface ContentfulArticle {
  type: string;
  title: string;
  abstract: string;
  seo_description: string;
  seo_keywords: string;
  slug: string;
  website: Entry<ContentfulWebsite>;
  published_at: string;
  authors: Entry<ContentfulAuthor>[];
  blocks: Entry<any>[];
  image_list: ContentfulAsset;
  image_header: ContentfulAsset;
}
