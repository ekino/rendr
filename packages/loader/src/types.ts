import { Page, RendrCtx, PageType } from "@ekino/rendr-core";
import { AxiosRequestConfig } from "axios";

export type MaybePage = Promise<PageType> | PageType;

export type Loader = (
  ctx: RendrCtx,
  page: PageType,
  next?: (page?: MaybePage) => MaybePage
) => MaybePage;

export type PageBuilder = (ctx: RendrCtx, page: Page) => MaybePage;

export type RouteConfiguration = {
  path: string;
  matcher: (path: string) => {} | undefined;
  pageBuilder: PageBuilder | Loader;
};

export interface InMemorySettings {
  [index: string]: PageBuilder | Loader;
}

export type AxiosOptionsBuilder = (
  url: string,
  options: AxiosRequestConfig
) => { url: string; options: AxiosRequestConfig };
