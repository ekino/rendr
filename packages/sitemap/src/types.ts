export type Options = {
  readonly basePathIndex: string;
};
