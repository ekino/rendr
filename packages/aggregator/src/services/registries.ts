import { HandlerList, HandlerRegistry } from "../types";
import { BlockDefinition } from "@ekino/rendr-core";

export function createHandlerRegistry(handlers: HandlerList): HandlerRegistry {
  return (code: string) => {
    if (code in handlers) {
      return handlers[code];
    }

    return (definition: BlockDefinition): Promise<BlockDefinition> => {
      // if no aggregator found for the block, we just return the block definition as is.
      return Promise.resolve(definition);
    };
  };
}
