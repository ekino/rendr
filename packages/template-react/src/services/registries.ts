import NotFoundComponent from "../components/NotFoundComponent";
import DefaultTemplate from "../templates/DefaultTemplate";

import { ComponentList, BlockRegistry, TemplateRegistry } from "../types";

export function createBlockRegistry(blocks: ComponentList): BlockRegistry {
  if (!("default" in blocks)) {
    blocks["default"] = NotFoundComponent;
  }

  return (code, settings) => {
    if (code in blocks) {
      return {
        component: blocks[code],
        settings,
      };
    }

    return {
      component: blocks["default"],
      settings: {
        name: code,
        settings,
      },
    };
  };
}

export function createTemplateRegistry(
  templates: ComponentList
): TemplateRegistry {
  if (!("default" in templates)) {
    templates["default"] = DefaultTemplate;
  }

  return (code) => {
    if (code in templates) {
      return templates[code];
    }

    return templates["default"];
  };
}
