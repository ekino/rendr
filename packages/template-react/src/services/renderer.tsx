import React from "react";
import { BlockRegistry, BlockRenderer, ContainerRenderer } from "../types";

export function createBlockRenderer(
  blockRegistry: BlockRegistry
): BlockRenderer {
  return function blockRenderer(block, key) {
    const { component: Component, settings: props } = blockRegistry(
      block.type,
      block.settings
    );

    return <Component key={key} {...props} blockRenderer={blockRenderer} />;
  };
}

export function createContainerRenderer(
  renderBlock: BlockRenderer
): ContainerRenderer {
  return (name, blocks) => {
    return (
      <>
        {blocks
          .filter(block => name === block.container)
          .map((block, i) => {
            return renderBlock(block, `k_${i}`);
          })}
      </>
    );
  };
}
