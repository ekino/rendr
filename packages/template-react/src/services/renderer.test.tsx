import React from "react";
import renderer from "react-test-renderer";
import { BlockDefinition } from "@ekino/rendr-core";

import { createBlockRenderer } from "./renderer";
import { createBlockRegistry } from "./registries";
import { BlockRendererProps } from "../types";

interface ArticleViewProps {
  blocks: BlockDefinition[];
}

const registry = createBlockRegistry({
  one: (props) => <div>One</div>,
  "article.view": (props: ArticleViewProps & BlockRendererProps) => {
    const { blocks, blockRenderer } = props;
    let id = 0;

    const blockComponents = blocks.map((block) =>
      blockRenderer(block, `key_${id++}`)
    );

    return <div>Two{blockComponents}</div>;
  },
});

const blockRenderer = createBlockRenderer(registry);

describe("Test renderer", () => {
  test("test simple component", () => {
    const block: BlockDefinition = {
      id: "",
      meta: {},
      type: "one",
      settings: {},
      container: "body",
      order: 1,
    };

    const output = blockRenderer(block, "key_1");

    const tree = renderer.create(output).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("test nested components", () => {
    const block: BlockDefinition = {
      type: "article.view",
      id: "",
      meta: {},
      settings: {
        blocks: [
          {
            type: "one",
            settings: {},
            container: "body",
            order: 1,
          },
          {
            type: "one",
            settings: {},
            container: "body",
            order: 1,
          },
        ],
      },
      container: "body",
      order: 1,
    };

    const output = blockRenderer(block, "key_2");

    const tree = renderer.create(output).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
