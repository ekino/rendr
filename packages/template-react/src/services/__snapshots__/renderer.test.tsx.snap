// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test renderer test nested components 1`] = `
<div>
  Two
  <div>
    One
  </div>
  <div>
    One
  </div>
</div>
`;

exports[`Test renderer test simple component 1`] = `
<div>
  One
</div>
`;
